lib/relations/Relation.js

```diff
@@ -148,6 +155,36 @@ Relation.prototype = {
         return fromPort !== toPort;
     },
 
+    get canonical() {
+        if (!('_canonical' in this)) {
+            var canonical = false;
+
+            if (this.assetGraph && this.assetGraph.canonicalRoot) {
+                var canonicalRootObj = urlModule.parse(this.assetGraph.canonicalRoot, false, true);
+                var hrefObj = urlModule.parse(this.href, false, true);
+
+                canonical = canonicalRootObj.host === hrefObj.host
+                    && hrefObj.path.indexOf(canonicalRootObj.path) === 0
+                    && (canonicalRootObj.protocol === hrefObj.protocol || canonicalRootObj.protocol === null);
+            }
+
+            this._canonical = canonical;
+        }
+
+        return this._canonical;
+    },
+
+    set canonical(isCanonical) {
+        this._canonical = !!isCanonical;
+
+        var hrefTypeRoot = this._canonical ? this.assetGraph.canonicalRoot : '';
+        // We need to update the hrefType manually here because otherwise hrefType in refreshHref
+        // will be based on the old href value, not the new one
+        this._hrefType = Object.getOwnPropertyDescriptor(Relation.prototype, 'hrefType').get.call({ href: hrefTypeRoot });
```

Comment:
Right, I see it now. Because "http://canonical.com/local.js" is an absolute url, the relation will get an `_hrefType` of `absolute` as soon as something accesses the `hrefType` getter. That remains unaffected by flipping `_canonical` to false, so `refreshHref` will issue an absolute `file://` url in that test.

What you want is to automatically switch to a "lesser" type of href if possible when the canonicalness is switched off so you won't end up with eg. an absolute `file://` url. I guess that makes sense -- but we don't really know if a `hrefType` of `relative` would be sufficient as the user hasn't really expressed the intent through that originally absolute href. Applications that use pushState generally need at least `rootRelative` hrefs.